File: python/problems/leetcode/0098_validate_binary_search_tree/solution.py
"""LeetCode 98: Validate Binary Search Tree."""

from __future__ import annotations

from math import inf
from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node: Optional[TreeNode], lo: float, hi: float) -> bool:
            if node is None:
                return True
            if not (lo < node.val < hi):
                return False
            return dfs(node.left, lo, node.val) and dfs(node.right, node.val, hi)

        return dfs(root, -inf, inf)


if __name__ == "__main__":
    # [2,1,3]
    assert Solution().isValidBST(TreeNode(2, TreeNode(1), TreeNode(3)))
    # [5,1,4,null,null,3,6]
    bad = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    assert not Solution().isValidBST(bad)
    print("0098 OK")

File: cpp/problems/leetcode/0098_validate_binary_search_tree/solution.cpp
// LeetCode 98
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, LLONG_MIN, LLONG_MAX);
    }

private:
    bool dfs(TreeNode* node, long long lo, long long hi) {
        if (!node) return true;
        if (node->val <= lo || node->val >= hi) return false;
        return dfs(node->left, lo, node->val) &&
               dfs(node->right, node->val, hi);
    }
};

int main() {
    TreeNode a(1), b(3), root(2, &a, &b);
    assert(Solution().isValidBST(&root));
    TreeNode c(3), d(6), e(4, &c, &d), f(1);
    TreeNode bad(5, &f, &e);
    assert(!Solution().isValidBST(&bad));
    cout << "0098 OK" << endl;
    return 0;
}
